File: e2e/php/tests/MetadataTest.php
<?php

declare(strict_types=1);

namespace Kreuzberg\E2e;

use PHPUnit\Framework\TestCase;
use Kreuzcrawl\Kreuzcrawl;

/** E2e tests for category: metadata. */
final class MetadataTest extends TestCase
{
    /** Extracts article:published_time, modified_time, author, section, and tags */
    public function test_metadata_article_times(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertEquals("2024-01-15T10:00:00Z", $result->article->published_time);
        $this->assertEquals("2024-06-20T14:30:00Z", $result->article->modified_time);
        $this->assertEquals("Jane Developer", $result->article->author);
        $this->assertEquals("Technology", $result->article->section);
        $this->assertEquals(3, count($result->article->tags));
    }

    /** Extracts favicon link tags including apple-touch-icon */
    public function test_metadata_favicons(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertEquals(5, count($result->favicons));
        $this->assertNotEmpty($result->favicons[""]->apple_touch);
    }

    /** Extracts heading hierarchy (h1-h6) from HTML page */
    public function test_metadata_headings(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertEquals(1, count($result->headings->h1));
        $this->assertEquals("Primary Heading", $result->headings->h1["0"]->text);
        $this->assertEquals(8, count($result->headings));
    }

    /** Extracts hreflang alternate link tags */
    public function test_metadata_hreflang(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertEquals(4, count($result->hreflang));
        $this->assertStringContainsString("en", $result->hreflang[""]->lang);
    }

    /** Extracts keywords, author, viewport, generator, theme-color, robots, lang, dir metadata */
    public function test_metadata_keywords_author(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertEquals("Comprehensive Metadata Test Page", $result->metadata->title);
        $this->assertNotEmpty($result->metadata->canonical_url);
        $this->assertNotEmpty($result->metadata->keywords);
        $this->assertStringContainsString("rust", $result->metadata->keywords);
        $this->assertEquals("Jane Developer", $result->metadata->author);
        $this->assertNotEmpty($result->metadata->viewport);
        $this->assertEquals("kreuzcrawl/1.0", $result->metadata->generator);
        $this->assertEquals("#ff6600", $result->metadata->theme_color);
        $this->assertEquals("index, follow", $result->metadata->robots);
        $this->assertEquals("en", $result->metadata->lang);
        $this->assertEquals("ltr", $result->metadata->dir);
    }

    /** Extracts og:video, og:audio, and og:locale:alternate metadata */
    public function test_metadata_og_video_audio(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertEquals("https://example.com/video.mp4", $result->og->video);
        $this->assertEquals("https://example.com/audio.mp3", $result->og->audio);
        $this->assertEquals(2, count($result->og->locale_alternate));
    }

    /** Extracts response metadata from HTTP headers (etag, server, content-language) */
    public function test_metadata_response_headers(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertNotEmpty($result->response_headers->etag);
        $this->assertNotEmpty($result->response_headers->last_modified);
        $this->assertStringContainsString("nginx", $result->response_headers->server);
        $this->assertEquals("en-US", $result->response_headers->content_language);
    }

    /** Computes word count from visible page text */
    public function test_metadata_word_count(): void
    {
        $engine = Kreuzcrawl::createEngine(null);
        $result = Kreuzcrawl::scrape($engine, "");
        $this->assertEquals(200, $result->status_code);
        $this->assertGreaterThan(99, $result->computed->word_count);
        $this->assertLessThan(301, $result->computed->word_count);
    }
}

File: e2e/rust/tests/metadata_test.rs
//! E2e tests for category: metadata

use kreuzcrawl::scrape;
use kreuzcrawl::create_engine;

#[test]
fn test_metadata_article_times() {
    // Extracts article:published_time, modified_time, author, section, and tags
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    // skipped: field 'article.published_time' not available on result type
    // skipped: field 'article.modified_time' not available on result type
    // skipped: field 'article.author' not available on result type
    // skipped: field 'article.section' not available on result type
    // skipped: field 'article.tags.length' not available on result type
}

#[test]
fn test_metadata_favicons() {
    // Extracts favicon link tags including apple-touch-icon
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    // skipped: field 'favicons.length' not available on result type
    // skipped: field 'favicons[].apple_touch' not available on result type
}

#[test]
fn test_metadata_headings() {
    // Extracts heading hierarchy (h1-h6) from HTML page
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    // skipped: field 'headings.h1.length' not available on result type
    // skipped: field 'headings.h1[0].text' not available on result type
    // skipped: field 'headings.length' not available on result type
}

#[test]
fn test_metadata_hreflang() {
    // Extracts hreflang alternate link tags
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    // skipped: field 'hreflang.length' not available on result type
    // skipped: field 'hreflang[].lang' not available on result type
}

#[test]
fn test_metadata_keywords_author() {
    // Extracts keywords, author, viewport, generator, theme-color, robots, lang, dir metadata
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    let metadata_title = result.metadata.title.as_deref().unwrap_or("");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    assert_eq!(metadata_title, r#"Comprehensive Metadata Test Page"#, "equals assertion failed");
    assert!(result.metadata.canonical_url.is_some(), "expected metadata.canonical_url to be present");
    assert!(!result.metadata.keywords.is_empty(), "expected non-empty value");
    assert!(result.metadata.keywords.contains(r#"rust"#), "expected to contain: {}", r#"rust"#);
    assert_eq!(result.metadata.author, r#"Jane Developer"#, "equals assertion failed");
    assert!(!result.metadata.viewport.is_empty(), "expected non-empty value");
    assert_eq!(result.metadata.generator, r#"kreuzcrawl/1.0"#, "equals assertion failed");
    assert_eq!(result.metadata.theme_color, r#"#ff6600"#, "equals assertion failed");
    assert_eq!(result.metadata.robots, r#"index, follow"#, "equals assertion failed");
    assert_eq!(result.metadata.lang, r#"en"#, "equals assertion failed");
    assert_eq!(result.metadata.dir, r#"ltr"#, "equals assertion failed");
}

#[test]
fn test_metadata_og_video_audio() {
    // Extracts og:video, og:audio, and og:locale:alternate metadata
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    // skipped: field 'og.video' not available on result type
    // skipped: field 'og.audio' not available on result type
    // skipped: field 'og.locale_alternate.length' not available on result type
}

#[test]
fn test_metadata_response_headers() {
    // Extracts response metadata from HTTP headers (etag, server, content-language)
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    // skipped: field 'response_headers.etag' not available on result type
    // skipped: field 'response_headers.last_modified' not available on result type
    // skipped: field 'response_headers.server' not available on result type
    // skipped: field 'response_headers.content_language' not available on result type
}

#[test]
fn test_metadata_word_count() {
    // Computes word count from visible page text
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, url).expect("should succeed");
    assert_eq!(result.status_code, "200", "equals assertion failed");
    // skipped: field 'computed.word_count' not available on result type
    // skipped: field 'computed.word_count' not available on result type
}

